File: py/rwshm.py
## @package rwshm - read write shared memory.
# `

from multiprocessing import Process
from multiprocessing.shared_memory import SharedMemory
from multiprocessing import shared_memory, resource_tracker
import numpy as np

#import telemetry
from telemetry import *

## rshm == read shared memory
class rshm:
    ## The constructor
    # @param self The object pointer
    # @param shmMapFilename A text string of the shared memory filename.
    # @param shmValueSize The shared memory size in bytes.
    def __init__(self, shmMapFilename, shmValueSize ):
        ## Member for SharedMemory object.
        self.shm = SharedMemory(name=shmMapFilename, create=False)
        ## Member for shared memory size in bytes.
        self.valueSize = shmValueSize
        ## Member for number of integers for shared memory.
        self.numInts = int(shmValueSize / 4)
        ## Member for array of integers mapped to shared memory.
        self.arr = np.ndarray([self.numInts], 'u4', buffer = self.shm.buf)

    ## Method to get read access to shared memory.
    # @param self The object pointer
    def getMapPtr(self):
        return self.arr;

    ## The destructor
    # @param self The object pointer
    def __del__(self):
        self.shm.close();
        self.shm.unlink();

## wshm == write shared memory.
class wshm:
    ## The constructor
    # @param self The object pointer
    # @param shmMapFilename A text string of the shared memory filename.
    # @param shmValueSize The shared memory size in bytes.
    def __init__(self, shmMapFilename='/shmem-example', shmValueSize=4 ):
        ## Member for SharedMemory object.
        self.shm = SharedMemory(size=shmValueSize*4, name=shmMapFilename, create=True)
        ## Member for shared memory size in bytes.
        self.valueSize = shmValueSize
        ## Member for number integers in shared memory.
        self.numInts = int( shmValueSize / 4 );
        ## Member array mapped to shared memory
        self.arr = np.ndarray([self.numInts], 'u4', buffer = self.shm.buf)
        for i in range(telemetry.kNumEntries):
           self.arr[i] = 0

    ## Method to get write access to shared memory.
    # @param self The object pointer
    def getMapPtr(self):
        return self.arr;

    ## The destructor
    # @param self The object pointer
    def __del__(self):
        self.shm.close();
        # The following statement needed so writing process can terminate
        # with shared memory in tact for later reading.
        resource_tracker.unregister(self.shm._name, 'shared_memory')
        #self.shm.unlink(); # Only unlink after reading.

## Writer for shared memory.
class writer(wshm):
    ## The constructor
    def __init__(self):
       wshm.__init__(self, shmMapFilename="/telemetry", shmValueSize=telemetry.kTotalByteSize)

    ## The destructor.
    # @param self The object pointer
    def __del__(self):
       wshm.__del__(self)

    ## Method for simple initialization for later verification.
    # @param self The object pointer
    def initIncPattern(self):
        for i in range(telemetry.kNumEntries):
           self.arr[i] = i

## Reader for reading telemetry data shared memory to compute bit counts.
class reader(rshm):
    ## The constructor
    def __init__(self):
       rshm.__init__(self, shmMapFilename="/telemetry", shmValueSize=telemetry.kTotalByteSize)
       self.clearTelemetry()

    ## The destructor
    # @param self The object pointer
    def __del__(self):
       rshm.__del__(self)

    ## Initialize the bit counts to zero.
    # @param self The object pointer
    def clearTelemetry( self ):
       ## Member array for storing bit counts
       self.bitCounts = np.zeros( [telemetry.kNumThreads,telemetry.kNumEntriesPerThread], dtype = int )

    ## Method to count 1 for every bit that is on for a sample(frame).
    # @param self The object pointer
    # @param sampleIdx The sample(frame) to analyze.
    def gatherTelemetryBitCounts( self, sampleIdx ):
       row = 0
       col = 0
       entryIdx = (sampleIdx * telemetry.kNumEntriesPerFrame)
       for entry in range(telemetry.kNumEntriesPerFrame):
          entryValue = self.arr[ entryIdx ]
          entryIdx += 1
          bitMask = 0x01
          count = 0;
          for bitIdx in range( 32 ):
             if ((entryValue & bitMask) != 0):
                count += 1
             bitMask = bitMask << 1
          self.bitCounts[row][col] += count
          col += 1
          if (col >= telemetry.kNumEntriesPerThread):
             col = 0
             row += 1

    ## Method to verify a simple pattern.
    # @param self The object pointer
    def verifyIncPattern( self ):
        RV = True
        for i in range(telemetry.kNumEntries):
            if (self.arr[i] != i):
                RV = False;
                break;
        return RV

    ## Print bit counts.
    # @param self The object pointer
    def printBitCounts( self ):
       for row in range(telemetry.kNumThreads):
          print( "ROW: %04x, " % row )
          line = "   "
          for col in range(telemetry.kNumEntriesPerThread):
             line += "%04x, " % self.bitCounts[row][col]
          print(line)

File: cpp/QtSln/utilLib/rwshm.cpp
//! @file rwshm.cpp
//! Method class details for reading and writing shared memory.
//! \author Andrew Louder
#include <iostream>
#include <iomanip>
#include "rwshm.h"

using namespace std;

wshm::wshm( const char* shmMapFile,
     size_t shmValueSz,
     int shmOpenFlag,
     int shmMapProt,
     mode_t shmOpenMode )
    : mapFile( shmMapFile ),
      valueSz(shmValueSz),
      openFlag(shmOpenFlag),
      mapProt(shmMapProt),
      openMode(shmOpenMode),
      mapDataPtr(NULL)
{
    fd = shm_open(mapFile, openFlag, openMode);
    if (fd < 0) {
        perror("shm_open()");
        exit( EXIT_FAILURE );
    }

    ftruncate(fd, valueSz);

    uint32_t *ptr = mapDataPtr = getMapPtr();

    for (uint32_t i=0; i < telemetry::kNumEntries; i++)
    {
        *ptr++ = 0;
    }
}

uint32_t * wshm::getMapPtr( void )
{
    if (mapDataPtr == NULL)
    {
        mapDataPtr = (uint32_t *)mmap(0, valueSz, mapProt, MAP_SHARED, fd, 0);
    }

    return mapDataPtr;
}

wshm::~wshm()
{
    cout << "wshm destructor called." << endl;
    if (mapDataPtr != NULL)
    {
        munmap(mapDataPtr, valueSz);
    }
    close(fd);
}

writer::writer() : wshm( "telemetry", telemetry::kTotalByteSize )
{

}

writer::~writer()
{
    cout << "writer destructor called" << endl;
}

void writer::initIncPattern( void )
{
    uint32_t *ptr = getMapPtr();

    for (uint32_t i=0; i < telemetry::kNumEntries; i++)
    {
        *ptr++ = i;
    }
}

void writer::initThreadEntry( uint8_t threadId, uint8_t entryIdx, uint32_t value )
{
    uint32_t *ptr = getMapPtr()
                    + (threadId * telemetry::kNumEntriesPerThread)
                    + (entryIdx);

    *ptr++ = value;
}

rshm::rshm( const char* shmMapFile,
     size_t shmValueSz,
     int shmOpenFlag,
     int shmMapProt,
     mode_t shmOpenMode )
    : mapFile( shmMapFile ),
      valueSz(shmValueSz),
      openFlag(shmOpenFlag),
      mapProt(shmMapProt),
      openMode(shmOpenMode)
{
    fd = shm_open(mapFile, openFlag, openMode);
    if (fd < 0) {
        perror("shm_open()");
        exit( EXIT_FAILURE );
    }
    mapDataPtr = NULL;
}

uint32_t * rshm::getMapPtr( void )
{
    if (mapDataPtr == NULL)
    {
        mapDataPtr = (uint32_t *)mmap(0, valueSz, PROT_READ, MAP_SHARED, fd, 0);
    }

    return mapDataPtr;
}

rshm::~rshm()
{
    cout << "rshm destructor called." << endl;
    close(fd);
    if (mapDataPtr != NULL)
    {
        munmap(mapDataPtr, valueSz);
    }
    shm_unlink( mapFile );
}

reader::reader() : rshm( "telemetry", telemetry::kTotalByteSize )
{

}
void reader::gatherTelemetryBitCounts( uint32_t numSamples )
{
    uint32_t *ptr = getMapPtr();
    uint32_t row;
    uint32_t col;
    uint32_t entryValue;
    uint32_t bitIdx;
    uint32_t bitMask;

    assert( numSamples < UINT16_MAX );

    for ( row=0; row < telemetry::kNumThreads; ++row)
    {
        for (col=0; col < telemetry::kNumEntriesPerThread; ++col)
        {
            bitCounts[row][col] = 0;
        }
    }
    for (uint32_t teleSample=0; teleSample < numSamples; ++teleSample)
    {
        row = 0;
        col = 0;
        for (uint32_t entry=0; entry < telemetry::kNumEntriesPerFrame; ++entry)
        {
            assert( row < telemetry::kNumThreads );
            assert( col < telemetry::kNumEntriesPerThread );
            entryValue = *ptr++;
            bitMask = 0x1;
            for (bitIdx=0; bitIdx < 32; bitIdx++)
            {
                if ((entryValue & bitMask) != 0)
                {
                    bitCounts[row][col]++;
                }
                bitMask = (bitMask << 1);
            }
            //cout << "r:" << row << ", c:" << col << ", E:" << entryValue << ", cnt:" << bitCounts[row][col] << endl;
            if (++col >= telemetry::kNumEntriesPerThread)
            {
                col = 0;
                row++;
            }
        }
        //printBitCounts();
    }
}

bool reader::verifyIncPattern( void )
{
    bool RV = true;
    uint32_t *ptr = getMapPtr();

    for ( uint32_t i=0; i < telemetry::kNumEntries; ++i)
    {
        if (*ptr++ != i)
        {
            RV = false;
            break;
        }
    }

    return RV;
}

void reader::printBitCounts( void )
{
    uint32_t row;
    uint32_t col;

    cout << "Bit counts" << endl;
    for ( row=0; row < telemetry::kNumThreads; ++row)
    {
        cout << hex << setfill('0') << setw(4) << "ROW: " << row << " : ";// << endl;
        for (col=0; col < telemetry::kNumEntriesPerThread; ++col)
        {
            cout << hex << setfill('0') << setw(4) << bitCounts[row][col] << ", ";
        }
        cout << endl;
    }
}

reader::~reader()
{

}
